tests/test_maker.py

```diff
@@ -0,0 +1,317 @@
+                color = (color_intensity, 100, 255 - color_intensity)
+                draw.rectangle(
+                    [disk_x1, disk_y, disk_x2, disk_y + disk_height - 5],
+                    fill=color,
+                    outline="black",
+                    width=2,
+                )
+        return img
+
+    def visualise(self):
+        img = self.visualise_image()
+        if img:
+            img.show()
+        else:
+            self.visualise_text()
+
+    def apply(self, step: Step) -> Optional["GameState"]:
+        """
+        Given a tower `start` and a target tower `end` moves the topmost disk to the end tower.
+        """
+        start, end = step.start, step.end
+
+        if not (0 <= start < len(self.towers) and 0 <= end < len(self.towers)):
+            return None
+
+        # start tower is non empty
+        if len(self.towers[start]) == 0:
+            return None
+
+        # end tower is a valid target
+        if len(self.towers[end]) > 0 and self.towers[start][-1] > self.towers[end][-1]:
+            return None
+
+        # create state with the move applied
+        new_towers = [list(tower) for tower in self.towers]
+        disk = new_towers[start].pop()
+        new_towers[end].append(disk)
+
+        #
+        new_state = GameState(
+            size=self.size, towers=tuple(tuple(tower) for tower in new_towers)
+        )
+        return new_state
+
+    def steps_to_complete(self) -> int:
+        """Compute the number of steps to complete the towers of hanoi from a given configuration if using the optimal algorithm."""
+        # Count disks on each tower
+        total_moves = 0
+
+        # For each tower that's not the destination, we need to move all its disks
+        for tower_idx, tower in enumerate(self.towers):
+            if tower_idx == self.size - 1:
+                continue
+
+            # Number of disks on this tower
+            n_disks = len(tower)
+
+            if n_disks > 0:
+                # Moving n disks from one peg to another requires 2^n - 1 moves
+                total_moves += (2**n_disks) - 1
+
+        return total_moves
+
+    def is_done(self) -> bool:
+        return all(len(tower) == 0 for tower in self.towers[:-1]) and all(
+            self.towers[-1][i] > self.towers[-1][i + 1]
+            for i in range(len(self.towers[-1]) - 1)
+        )
+
+    def valid_steps(self) -> list[Step]:
+        steps = []
+        for i, tower_i in enumerate(self.towers):
+            for j, tower_j in enumerate(self.towers):
+                if i == j:
+                    continue
+                if len(tower_i) == 0:
+                    continue
+                # if tower_i's disk is smaller than tower_j's topmost, then it is valid to move from tower i to j
+                if len(tower_j) == 0 or tower_i[-1] < tower_j[-1]:
+                    steps.append(Step(i, j))
+        return steps
+
+
+class MicroAgent:
```

Comment:
It seems like `MicroAgent` should just be a single stateless `Template` with a structured output type annotation?
```py
@Template.define
def predict_next_state(state: GameState) -> Step:
    """..."""
```
Here the behavior of `parse_response`, `has_no_red_flags` and `get_vote` should all be implicit in the annotation.

Reply:
`parse_response` should just be the default behavior of structured output generation. `has_no_red_flags` would follow either from a more specific static `Step` type or making it a Pydantic model with the runtime validation logic from `has_no_red_flags`.